portindex2json/portindex2json.tcl

```diff
@@ -64,7 +67,67 @@ proc remove_extra_braces {text} {
     return [regsub -all {[\{\}]} $text ""]
 }
 
-set fd [open [lindex $argv 0] r]
+proc print_usage {} {
+    puts stdout ""
+    puts stdout "Usage:"
+    puts stdout "    /path/to/tclsh portindex2json.tcl \[--info <key>=<value>\] <path/to/PortIndex>"
+    puts stdout "    /path/to/tclsh portindex2json.tcl --help"
+    puts stdout ""
+    puts stdout "Each '--info <key>=<value>' combinaton adds one key-value pair to the JSON output inside the parent key: \"info\""
+    puts stdout "You may provide any number of '--info <key>=<value>' combinations."
+    puts stdout ""
+}
+
+proc parse_options {} {
+    global argc argv fd
+    for {set i 0} {$i < $argc} {incr i} {
+        set arg [lindex $argv $i]
+        switch -regex -- $arg {
+            {^-.+} {
+                if {$arg eq "--info"} {
+                    incr i
+                    set argument [split [lindex $argv $i] =]
+                    if {[lindex $argument 1] eq ""} {
+                        puts stderr "\nERROR: No value provided for the key: '[lindex $argument 0]'. The correct syntax is '--info <key>=<value>'.\n"
+                        exit 1
+                    } else {
+                        set json_info([lindex $argument 0]) [::json::write string [lindex $argument 1]]
+                    }
+                } elseif {$arg eq "-help" || $arg eq "-h" || $arg eq "--help" } {
+                    print_usage
+                    exit 0
+                } else {
+                    puts stderr "\nERROR: '$arg' is an invalid argument (expecting '--help' or '--info').\n"
+                    print_usage
+                    exit 1
+                }
+            }
+            default {
+                if {[info exists fd]} {
+                    puts stderr "\nERROR: No support for more than one argument for path to the PortIndex file.\n"
+                    print_usage
+                    exit 1
+                } else {
+                    if {[catch {open [lindex $argv $i] r} fid]} {
+                        puts stderr "\nERROR: Could not open file '[lindex $argv $i]'. Make sure the path to the PortIndex file is correct.\n"
+                        exit 1
+                    } else {
+                        set fd [open [lindex $argv $i] r]
```

Comment:
Disclaimer: I'm hardly literate in Tcl.

While this obviously does the job, the part starting with `[catch {open ...}]` seems a bit suspicious. Do we ever close the file at all? Do we really need to call `open` twice?

@jmroot

Reply:
Oh yes, we can actually move the entire `set fd` statement inside `catch`. Thank you

Reply:
Or maybe try-catch?

Reply:
> Or maybe try-catch?

Does that give any other benefit. Just `catch` would have shorter syntax.

Reply:
I have added the commit to do this. Please complain if you still want to prefer `try-catch`.

Reply:
No, that's fine. I keep forgetting how concise the syntax is after working with a much more arcane language at work.